components/family-studies/migrations/src/main/java/org/phenotips/studies/family/migrations/R71507PhenoTips3423DataMigration.java

```diff
@@ -0,0 +1,558 @@
+                if (xDocument == null) {
+                    continue;
+                }
+
+                pedigreeXObject = xDocument.getXObject(PEDIGREE_CLASS_REFERENCE);
+                if (pedigreeXObject == null) {
+                    continue;
+                }
+            } catch (Exception e) {
+                this.logger.error("Error checking pedigree data for document {}: [{}]",
+                        docName, e.getMessage());
+                continue;
+            }
+
+            try {
+                this.logger.debug("Updating pedigree for document {}.", docName);
+
+                if (!this.updatePedigree(pedigreeXObject, context, docName)) {
+                    continue;
+                }
+
+                xDocument.setComment(this.getDescription());
+                xDocument.setMinorEdit(true);
+
+            } catch (Exception e) {
+                this.logger.error("Error updating pedigree data format for document {}: [{}]",
+                        docName, e.getMessage());
+                continue;
+            }
+
+            try {
+                // There's a bug in XWiki which prevents saving an object in the same session that it was loaded,
+                // so we must clear the session cache first.
+                session.clear();
+                ((XWikiHibernateStore) getStore()).saveXWikiDoc(xDocument, context, false);
+                session.flush();
+            } catch (DataMigrationException e) {
+                this.logger.error("Error when saving XWiki document {}: [{}]", docName, e.getMessage());
+            }
+
+        }
+        return null;
+    }
+
+    // used to store data about a (potentially multi-generation) path (in a graph sense)
+    // from partners to their relationships
+    private static class PartnerPath
+    {
+        private Integer parentID;
+        private List<Integer> path;
+
+        PartnerPath(Integer parentID)
+        {
+            this.parentID = parentID;
+            this.path = new LinkedList<>();
+        }
+
+        public void addPathElement(Integer virtualNodeID)
+        {
+            this.path.add(virtualNodeID);
+        }
+
+        public Integer getParentID()
+        {
+            return this.parentID;
+        }
+
+        public List<Integer> getPath()
+        {
+            return this.path;
+        }
+
+        public String toString()
+        {
+            if (this.path.size() == 0) {
+                return this.parentID.toString() + " (direct)";
+            } else {
+                return this.parentID.toString() + " -> <" + this.path.toString() + ">";
+            }
+        }
+    }
+
+    private boolean updatePedigree(BaseObject pedigreeXObject, XWikiContext context, String documentName)
+    {
+        String oldPedigreeAsText = pedigreeXObject.getStringValue(PEDIGREECLASS_JSONDATA_KEY);
+        if (!StringUtils.isEmpty(oldPedigreeAsText)) {
```

Comment:
why not `StringUtils.isBlank()`

Reply:
Because "empty" means no data (which is OK, just "do nothing"), while blank (but not empty) means corrupt data - and it is best to handle corrupt data together with other cases of corrupt data e.g. non-blank, non-empty, but still wrong

Reply:
(to be precise, empty is also "not OK" here, but it is still a different kind of error compared to wrong data)